src/http/gateway/resolver.js

```diff
@@ -0,0 +1,112 @@
+'use strict'
+
+const mh = require('multihashes')
+const promisify = require('promisify-es6')
+const eachOfSeries = require('async/eachOfSeries')
+const CID = require('cids')
+const Unixfs = require('ipfs-unixfs')
+const debug = require('debug')
+const log = debug('jsipfs:http-gateway:resolver')
+log.error = debug('jsipfs:http-gateway:resolver:error')
+
+const dirView = require('./dir-view')
+const pathUtil = require('./utils/path')
+
+function getIndexFiles (links) {
+  const INDEX_HTML_FILES = [
+    'index.html',
+    'index.htm',
+    'index.shtml'
+  ]
+
+  return links.filter((link) => INDEX_HTML_FILES.indexOf(link.name) !== -1)
+}
+
+function noop () {}
+
+const resolveDirectory = promisify((ipfs, path, multihash, callback) => {
+  callback = callback || noop
+
+  mh.validate(mh.fromB58String(multihash))
+
+  ipfs.object.get(multihash, { enc: 'base58' }, (err, dagNode) => {
+    if (err) { return callback(err) }
+
+    const indexFiles = getIndexFiles(dagNode.links)
+
+    if (indexFiles.length > 0) {
+      return callback(null, indexFiles)
+    }
+
+    return callback(null, dirView.render(path, dagNode.links))
+  })
+})
+
+const resolveMultihash = promisify((ipfs, path, callback) => {
+  callback = callback || noop
+
+  const parts = pathUtil.splitPath(path)
+  const partsLength = parts.length
+
+  let currentMultihash = parts[0]
+  let currentCid
+  eachOfSeries(parts, (multihash, currentIndex, next) => {
+    // throws error when invalid CID is passed
+    try {
+      currentCid = new CID(mh.fromB58String(currentMultihash))
+    } catch (err) {
+      return next(err)
+    }
+
+    log('currentMultihash: ', currentMultihash)
+    log('currentIndex: ', currentIndex, '/', partsLength)
+
+    ipfs.dag.get(currentCid, (err, result) => {
+      if (err) { return next(err) }
+      let dagNode = result.value
+
+      if (currentIndex === partsLength - 1) {
+        let dagDataObj = Unixfs.unmarshal(dagNode.data)
+        if (dagDataObj.type === 'directory') {
+          let isDirErr = new Error('This dag node is a directory')
+          // add currentMultihash as a fileName so it can be used by resolveDirectory
+          isDirErr.fileName = currentMultihash
+          return next(isDirErr)
+        }
+
+        return next()
+      }
+
+      // find multihash of requested named-file in current dagNode's links
+      let multihashOfNextFile
+      const nextFileName = parts[currentIndex + 1]
+      const links = dagNode.links
+
+      for (let link of links) {
+        if (link.name === nextFileName) {
+          // found multihash of requested named-file
+          multihashOfNextFile = mh.toB58String(link.multihash)
+          log('found multihash: ', multihashOfNextFile)
+          break
+        }
+      }
+
+      if (!multihashOfNextFile) {
+        log.error(`no link named "${nextFileName}" under ${currentMultihash}`)
+        return next(new Error(`no link named "${nextFileName}" under ${currentMultihash}`))
```

Comment:
Logging plus returning seems duplicate, shouldn't the returned error always be logged already?